README.md
# bedrock-payment-http


> Provides the HTTP APIs for bedrock-payment.

Inserts a set of API endpoints into a bedrock server that can securely process payments.

## Table of Contents

- [Security](#security)
- [Background](#background)
- [Install](#install)
- [Usage](#usage)
- [API](#api)
- [Maintainers](#maintainers)
- [Contributing](#contributing)
- [License](#license)

## Security
Bedrock Payment HTTP provides API endpoints for payments that require a user is logged in. Additional checks are made to ensure that the payment being processed was created by the account in the current session.

## Background
Bedrock Payment HTTP was created in order to provide an easy to use and secure API for payments from multiple payment gateways (i.e. paypal, stripe, etc.).

## Install

```sh
npm install bedrock-payment-http --save
```

## Usage
You can configure bedrock-payment-http by requiring a config file like this

```js
const bedrock = require('bedrock');
const path = require('path');

const {config} = bedrock;
const cfg = config['payment-http'];

const basePath = '/payment';

cfg.routes = {
  basePath
};
```

## API
This module adds several new API endpoints to your bedrock server.
the basePath defaults to /payment

### GET /payment/credentials?service=foo

Gets the credentials needed for client side charges.

  * Response codes:
     * 200: Credentials were found.
     * 404: Credentials not found.
     * 500: Server failure.


### GET /payment

Gets all payments for the currently logged in account.

  * Response codes:
     * 200: Request succeeded.
     * 500: Server failure.

### POST /payment

Creates or updates a pending payment.

  * Response codes:
     * 200 Payment Update Request succeeded.
     * 201 Payment Create Request succeeded.
     * 404 Payment not found.
     * 500 Server failure.

### PUT /payment/:paymentID

Processes a Payment.

  * Response codes:
    * 200 Payment Process succeeded.
    * 404 Payment not found.
    * 500 Server failure.

## Maintainers

[@DigitalBazaar](https://github.com/DigitalBazaar)

## Contributing

See [the Bedrock contributing file](https://github.com/digitalbazaar/bedrock/blob/master/CONTRIBUTING.md)!

PRs accepted.

Small note: If editing the README, please conform to the [standard-readme](https://github.com/RichardLitt/standard-readme) specification.

## License

Bedrock License © 2019 Digital Bazaar, Inc.


CHANGELOG.md
## 0.0.1 -

### Added
- lib/schemas/payment.js with validators for processing and creating payments.
- lib/index.js which sets up the configurable payment endpoints.
- Adds a config module that defaults the basePath to /payment.
- Adds a route GET /payment/credentials for payment api keys / auth.
- Adds a route GET /payment for getting all payments.
- Adds a route POST /payment for creating a new payment.
- Adds a route POST /payment/:paymentId for processing a new payment. 


lib/schemas/payment.js
/**!
 * Copyright (c) 2019 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';

const {config} = require('bedrock');
const {PaymentStatus} = require('bedrock-payment');
const {schemas} = require('bedrock-validation');
const currencyEnum = require('./currencies.json');

const numericFloat = '^[0-9]+(\\.[0-9]{0,2})?$';
const paymentEnum = Object.values(PaymentStatus);
// this is used by the validator to ensure you
// can not post a service not supported by the current project.
const servicesEnum = config.bedrock_payment.services || [];

const _payment = ({required}) => ({
  title: 'Bedrock Payment',
  type: 'object',
  required,
  additionalProperties: false,
  properties: {
    id: {
      title: 'id',
      type: 'string',
    },
    amount: {
      title: 'amount',
      type: 'string',
      pattern: numericFloat
    },
    currency: {
      title: 'currency',
      type: 'string',
      enum: currencyEnum
    },
    creator: {
      title: 'creator',
      type: 'string'
    },
    validated: {
      title: 'validated',
      anyOf: [{type: 'boolean'}, {type: 'null'}]
    },
    paymentService: {
      title: 'service',
      name: 'service',
      type: 'string',
      enum: servicesEnum
    },
    paymentServiceId: {
      title: 'serviceId',
      type: 'string'
    },
    status: {
      title: 'status',
      type: 'string',
      enum: paymentEnum
    },
    error: {
      title: 'error',
      anyOf: [{type: 'object'}, {type: 'null'}]
    },
    orderId: {
      title: 'orderId',
      type: 'string',
      minItems: 1
    },
    orderService: {
      title: 'orderService',
      type: 'string'
    },
    created: schemas.w3cDateTime()
  }
});

const all = [
  'id', 'amount', 'currency', 'creator', 'validated', 'paymentService',
  'paymentServiceId', 'status', 'error', 'orderId', 'orderService', 'created'
];

const creating = () => ({
  type: 'object',
  required: ['payment'],
  additionalProperties: false,
  properties: {
    payment: _payment({required: [
      'amount', 'paymentService',
      'orderId', 'orderService'
    ]})
  }
});

const processing = () => ({
  type: 'object',
  // order is optional and make contain gateway specific details.
  required: ['payment'],
  additionalProperties: false,
  properties: {
    payment: _payment({required: all}),
    order: {
      title: 'order',
      anyOf: [{type: 'object'}, {type: 'array'}, {type: 'null'}]
    }
  }
});

module.exports = {creating, processing};


lib/index.js
/**!
 * Copyright (c) 2019 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';

const bedrock = require('bedrock');
const brPassport = require('bedrock-passport');
const {validate} = require('bedrock-validation');
const {asyncHandler} = require('bedrock-express');
const paymentService = require('bedrock-payment');
const logger = require('./logger');

require('./config.js');

const {PaymentStatus, Errors} = paymentService;

const {ensureAuthenticated} = brPassport;
const {config} = bedrock;
const {BedrockError} = bedrock.util;

bedrock.events.on('bedrock-express.configure.routes', app => {
  const cfg = config['payment-http'];
  const {basePath: PAYMENT_ENDPOINT} = cfg.routes;
  // GET /payment/credentials
  // GET the api keys/auth tokens necessary to implement charges client side.
  app.get(
    `${PAYMENT_ENDPOINT}/credentials`,
    ensureAuthenticated,
    asyncHandler(async (req, res) => {
      const {service} = req.query;
      if(!service) {
        throw new BedrockError(
          'You must provide a service in the query.', Errors.Syntax);
      }
      // this will throw if the credential is not found.
      const data = await paymentService.getGatewayCredentials({service});
      return res.status(200).json(data);
    })
  );

  // GET /payment
  // GET all payments for an account.
  app.get(
    PAYMENT_ENDPOINT,
    ensureAuthenticated,
    asyncHandler(async (req, res) => {
      // this is so a authenticated account can not request
      // another account's payments.
      const {id: creator} = req.user.account;
      // check for existing Payments.
      const query = {creator};
      const payments = await paymentService.db.findAll({query});
      return res.status(200).json(payments);
    })
  );

  // POST /payment
  // Creates a new Payment.
  // The client will need to post the payment they intend to use.
  app.post(
    PAYMENT_ENDPOINT,
    ensureAuthenticated,
    validate({body: 'payment.creating'}),
    asyncHandler(async (req, res) => {
      const {id: creator} = req.user.account;
      const {payment: paymentData} = req.body;
      logger.debug('create payment called', {paymentData});
      const processingPayments = await paymentService.db.findAll(
        {query: {creator, status: PaymentStatus.PROCESSING}});
      // this can cause issues if a payment fails while processing.
      if(processingPayments.length > 0) {
        throw new BedrockError(
          'Can not create a new payment if you have processing payments.',
          Error.Constraint, {httpStatusCode: 409, public: true});
      }
      // check for existing pending Payments.
      const pendingPayments = await paymentService.db.findAll(
        {query: {creator, status: PaymentStatus.PENDING}});
      if(pendingPayments.length > 0) {
        logger.debug('updating payment', {creator, paymentData});
        // this is where most of the concurrency issues appear.
        const {updatedOrder, payment} = await paymentService.
          updatePendingPayment({pendingPayments, updatedPayment: paymentData});
        return res.status(200).json({order: updatedOrder, payment});
      }
      logger.debug('creating new payment', {creator, paymentData});
      const {order, payment} = await paymentService.createPayment(
        {creator, paymentData});
      return res.status(201).json({order, payment});
    }));

  // PUT /payment/:paymentId.
  // Processes an existing payment.
  app.put(
    `${PAYMENT_ENDPOINT}/:paymentId`,
    ensureAuthenticated,
    validate({body: 'payment.processing'}),
    asyncHandler(async (req, res) => {
      logger.debug('process payment called', {body: req.body});
      const {id: creator} = req.user.account;
      const {payment: paymentData, order = null} = req.body;
      // get the payment
      const payment = await paymentService.db.findOne(
        {query: {id: paymentData.id, creator}});
      if(!payment) {
        throw new BedrockError(
          'Payment not found',
          Errors.NotFound,
          {httpStatusCode: 404, public: true});
      }
      // this throws a BedrockError if the payment is already finished.
      // It has to be called before we make any change to the payment status.
      paymentService.finished({payment});
      payment.status = PaymentStatus.PROCESSING;
      await paymentService.db.save({payment});
      const verifiedPurchase = await paymentService.process({payment, order});
      logger.debug('verifiedPurchase', {verifiedPurchase});
      // this will throw an error if the verifiedPurchase
      // is not equal to the plan's amount.
      // it will also throw if the orderProcessor name is set incorrectly.
      const orderService = paymentService.order.get({id: payment.orderService});
      const orderConfirmed = await orderService.process(
        {payment, verifiedPurchase, order});
      return res.status(200).json(orderConfirmed);
    })
  );

});
